Improved warnings on closing projects with unsaved layers and memory layers

nyalldawson opened the pull request:
This PR:

1. Improves the current "unsaved changes in a layer" warnings on closing projects, to move this warning to occur BEFORE the "unsaved project" dialog. This now allows users to cancel the close when a layer has unsaved edits, where before they had no option but to either discard or save the changes. Sometimes you want to manually see and decide what action to take - which the new cancel option allows. It also fixes a bug where sometimes the "unsaved layer changes" warning wouldn't appear for some layers. Additionally, we now avoid showing the "unsaved layer changes" warning for memory layers -- it's misleading, because on project close clicking "save" to this warning has no effect -- the memory layer's contents are discarded anyway.

2. Adds a new warning on project close if a layer contains ANY memory layers with features present. These layers are temporary only, and their contents will be lost on project close. Again, the default action is to cancel the project close, allowing users the chance to manually save these layers out to a permanent location if desired.

Should help avoid data loss for users working with memory layers, especially memory layers created as a result of processing algorithms. And it'll pair very nicely with #7516!

Note that I believe this should be classified as a bug fix and intend to backport to 3.2.

Screencast:

![peek 2018-08-03 11-28](https://user-images.githubusercontent.com/1829991/43619414-fecbd51e-9710-11e8-869e-9f71a04048bc.gif)

nyalldawson committed:
Improved UX when closing projects with layers with unsaved changes

Previously, the unsaved layer warnings would come AFTER the
unsaved project warning. The unsaved layer warnings had no
cancel option, which meant that there was no way to cancel
a project close when unsaved layer edits exist, and you
were forced to either save these edits or discard them.

Now, the unsaved layer edits checks occur before the unsaved
project check, and you have the option now to 'Cancel' the
project close from the unsaved layer edits messagebox. This
allows users to cancel the project close operation if they
want to e.g. inspect manually unsaved layers to decide
what action is appropriate to take on these.

Also fixes the unsaved layer edits prompt sometimes didn't
appear, even though unsaved changes were present...

nyalldawson committed:
Warn users on closing projects with populated memory layers present

Adds a warning (with a cancel option!) when users start to close
a project containing any memory layers with features. These
layers are temporary only and their contents will be permanently
lost if the project is closed. The warning allows users to
cancel the close operation (that's the default action!) so that
they can then save these layers out to a permanent location.

nyalldawson commented:
Relates to https://issues.qgis.org/issues/19038 , https://issues.qgis.org/issues/18134, https://issues.qgis.org/issues/8387

nirvn commented:
Great.

One point: I think we need an (unexposed through UI) setting to switch this warning off. That would at least come in handy for the memory layer saver plugin.

Gustry commented:
Great, thanks a lot again.
Just a question, why not checking layers which are stored in the OS temporary directory? (/tmp on mac/linux) I think a few plugins do that (InaSAFE for example by default if you don't specify an output result, QuickOSM too in QGIS 2)

nyalldawson committed:
Add hidden setting to skip the project memory layer warning

To be used by 3rd party plugins, e.g. memory layer saver, who
may want to surpress this warning.

nyalldawson commented:
@Gustry 

That's a good idea, but it would need to be left to a follow-up piece of work. At the moment there's no direct method to retrieve the file path of a map layer (which is a major API short coming and needs fixing!). When that's added it should be trivial to make this check also for /tmp/ etc.

nyalldawson commented:
@nirvn 

> One point: I think we need an (unexposed through UI) setting to switch this warning off. That would at least come in handy for the memory layer saver plugin.

Done

nirvn commented:
Thanks. +1 to merge.

nirvn commented:
Also agreed it should be backported to 3.2

nyalldawson commented:
Thanks for the review @nirvn !

Gustry commented:
@nyalldawson Is-it normal that I don't have any warnings when I have some memory layers and I try to save the project (ctrl+s) without closing the project for instance.

It's also the same workflow detailed in the issue: https://issues.qgis.org/issues/19038

```
    Start QGIS 3
    Menu: Project > New
    Menu: Web > QuickMapServices > OSM > OSM Standart [optional step for OSM background tiles]
    Menu: Vector > Quick OSM > QickOSM
    Set query parameters to: {key: leisure, value: park, in: Leipzig}
    Run query ...
    Result so far: 3 new layers (with 599, 98 and 128 objects)
    Menu: Project > Save as ... "test.qgs"
    Menu: Project > Quit QGIS
    At this point, QGIS will ask: "Save current project?" I chose "Save".
    Start QGIS again
    Open "test.qgs" - all three new layers exist, but they contain ZERO objects
```

I got the warning after the "save as" when I close QGIS. But it's already too late, I saved my project with empty layers inside.

nyalldawson commented:
> @nyalldawson Is-it normal that I don't have any warnings when I have some memory layers and I try to save the project (ctrl+s) without closing the project for instance.

Yes - that's expected. The warning is triggered on project close, not project save.
